### ISSUE TITLE
Wrong (unwanted) partition in multiple-partition-range scan

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#9659]
I'm seeing this bug in the context of Alternator's expired-item scanner, but I suspect it is a much more fundamental bug on what happens when a read request includes multiple `partition_ranges`.

In one test, Alternator's scanner asked to read the following five ranges in one read request

        {[{-9040976251946270720, start}, {-9038724452132585472, start}],
         [{-9036472652318900224, start}, {-9034220852505214976, start}],
         [{-9031969052691529728, start}, {-9029717252877844480, start}],
         [{-9027465453064159232, start}, {-9025213653250473984, start}],
         [{-9022961853436788736, start}, {-9020710053623103488, start}]}

As you can see, the token **-9033214303902456950** is **not** in any of those ranges - it falls in the gap between the second and third ranges. But a partition with that token was returned, incorrectly, by this query!

I tried to debug this failure, and found the following suspicious code in `storage_proxy::query_partition_key_range_concurrent` written six years ago by @gleb-cloudius 
```c++
            // If we get there, merge this range and the next one
            range = dht::partition_range(range.start(), next_range.end());
```

Unfortunately I don't know what this is supposed to be doing, but when I log what this does I see in the log
```
WARN  2021-11-18 15:20:55,653 [shard 1] storage_proxy - merging [{-9036472652318900224, start}, {-9034220852505214976, start}] with [{-9031969052691529728, start}, {-9029717252877844480, start}]
``` 
And if we "merge" these two ranges with the gap between them into one range without a gap, the merged range suddenly contains the gap - and the item the query didn't ask for!!

It's worth noting the five ranges in this read request aren't random ranges - they were chosen specifically because they belong to the same shard of the same node. Perhaps this is what mislead us to think we can merge the two ranges? Yet if we do merge the two ranges it's critical we only read the merge range from that specific shard which doesn't have data in the gap, because other shards do have data in that gap.

I'm worried that this bug might have implications not just in the Alternator expiration scanner prototype, but also in CQL queries. I'll try to think how to reproduce this bug with CQL - although it will be difficult without understanding the purpose of this "merging" and why it is done. @gleb-cloudius or anyone else familiar with this code (@denesb ? @tgrabiec ?) can you perhaps help me by saying a few words about the meaning of this "merging" and why it's not buggy (or, maybe it is)?

CC @eliransin

**Comment by psarna** [comment]
Here's a whole discussion related to this problem, it happened to come up when coding the "distribute count(*)" project: https://github.com/scylladb/scylla/pull/9175

**Comment by psarna** [comment]
tldr: this code was never meant to be run on disjoint ranges, but the assumption is not obvious; it wasn't problematic before because of the reason you already mentioned - classic CQL does not have an OR operator

**Comment by denesb** [comment]
This merging is an optimization to reduce the amount of requests this code sends out to nodes. By default we have 1 request/1 range, but if we find that two neighboring ranges have an intersecting replica set > CL, then we merge the two ranges. Note that like @psarna said this assumes no holes between ranges, which is true for all ranges coming from CQL. The ranges list here is currently always obtained by chopping up one big continuous  into per-vnode ranges.

**Comment by psarna** [comment]
@havaker what's the status of https://github.com/scylladb/scylla/pull/9175 - was it decided that we don't want it or did it just die in review? If the latter, perhaps it needs a ping, especially now that we have another interested party.

**Comment by havaker** [comment]
it just died

**Comment by nyh** [comment]
I'm closing this issue as a duplicate of #9167.
